src/charm.py

```diff
@@ -73,13 +73,18 @@ def _on_install(self, _: ops.InstallEvent):
     def _on_start(self, event: ops.StartEvent):
         """Handle start event."""
         if not self.context.peer_relation:
-            self.unit.status = ops.WaitingStatus("waiting for peer relation")
+            self._set_status(Status.NO_PEER_RELATION)
             event.defer()
             return
 
-        if self.unit.is_leader() and not self.context.cluster.internal_user_credentials:
-            logger.info("Creating internal user")
+        if self.context.cluster.internal_user_credentials:
+            self.auth_manager.update_admin_user()
+        elif self.unit.is_leader():
             self.auth_manager.create_internal_user()
+        else:
+            # Unit is not leader and there are no internal credentials added yet
+            event.defer()
+            return
```

Comment:
**question:** This logic seems odd to me, but I think I might be misunderstanding, so I'll lay out what I think and you can correct me hopefully.

- `on_start`, we want to `update_admin_user`
- the first time I get `start`, if I'm the leader, I create an internal user and write to relation data
- the first time I get `start`, if I'm NOT the leader, I check if the internal user exists, otherwise I defer

If that's correct, I would have expected the flow to be:

```python
# early exit, to be ran on all units, including leader
if self.context.cluster.internal_user_credentials:
    self.auth_manager.update_admin_user()
    return

if self.unit.is_leader():
    self.auth_manager.create_internal_user()

# all units defer, so as to eventually run `update_admin_user()`
event.defer()

Reply:
I think your suggested code and mine are doing the same. Maybe the naming of the functions are misleading here, but `create_interal_user` is already updating the file as well, not just dealing with the databag, so leader will never need to `defer()` or run this code again.